Listen on loopback address during tests

casey opened the pull request:
This is a bit weird, since there's no good way to test that we're
actually listening on 0.0.0.0.

soenkehahn reviewed (approved):
I would've thought we would pass in e.g. `--address 127.0.0.1` during tests. (And implement `--address` first.) But I think this is also fine, since we do have a smoke test.

casey commented:
That sounds fine to me. I implemented `--address`. Looking up an address is somewhat complicated, because `ToSocketAddrs` does host resolution, and so might produce an I/O error. Also, the OS might return non-IP addresses, which are ignored, so the resulting iterator might be empty, which also needs to be handled.

I added a test that checks that the right error is produced with the right hostname.

soenkehahn commented on src/server.rs:
```diff
@@ -43,7 +53,7 @@ mod tests {
 
   #[test]
-  fn listen_on_all_local_ip_addresses() {
+  fn listen_on_localhost() {
```
```suggestion
  fn listen_on_localhost_by_default_in_tests() {
```

soenkehahn commented on tests/smoke.rs:
```diff
@@ -7,7 +7,7 @@ use std::{
 
 #[test]
-fn server_listens_on_localhost_8080() {
+fn server_listens_on_all_ip_addresses_port_8080() {
```
```suggestion
fn server_listens_on_all_ip_addresses_on_port_8080() {
```
